pymc_experimental/inference/smc/sampling.py

```diff
@@ -0,0 +1,421 @@
+#   Copyright 2023 The PyMC Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import logging
+import time
+import warnings
+from typing import Callable, Dict, NamedTuple, Optional, cast
+
+import arviz as az
+import blackjax
+import jax
+import jax.numpy as jnp
+import numpy as np
+from blackjax.smc.resampling import systematic
+from pymc import draw, modelcontext, to_inference_data
+from pymc.backends import NDArray
+from pymc.backends.base import MultiTrace
+from pymc.initial_point import make_initial_point_expression
+from pymc.sampling.jax import get_jaxified_graph
+from pymc.util import RandomState, _get_seeds_per_chain
+
+log = logging.getLogger(__name__)
+
+
+def sample_with_blackjax_smc(
+    n_particles,
+    random_seed: RandomState = None,
+    kernel: str = "HMC",
+    target_ess: float = 0.8,
+    num_mcmc_steps: int = 10,
+    inner_kernel_params: Optional[dict] = None,
+    model=None,
+    iterations_to_diagnose: int = 100,
+):
+    """Samples using BlackJax's implementation of Sequential Monte Carlo.
+       A summary of the algorithm is:
+
+     1. Initialize :math:`\beta` at zero and stage at zero.
+     2. Generate N samples :math:`S_{\beta}` from the prior (because when :math `\beta = 0` the
+        tempered posterior is the prior).
+     3. Increase :math:`\beta` in order to make the effective sample size equal some predefined
+        value (target_ess)
+     4. Compute a set of N importance weights W. The weights are computed as the ratio of the
+        likelihoods of a sample at stage i+1 and stage i.
+     5. Obtain :math:`S_{w}` by re-sampling according to W.
+     6. Run N independent MCMC chains, starting each one from a different sample
+        in :math:`S_{w}`. For that, set the kernel and inner_kernel_params.
+     7. The N chains are run for num_mcmc_steps each.
+     8. Repeat from step 3 until :math:`\beta \\ge 1`.
+     9. The final result is a collection of N samples from the posterior
+
+    Parameters
+    ----------
+    n_particles
+    random_seed
+    kernel
+    target_ess
```

Comment:
nitpick: maybe change the name to target_essn because is not the ESS but a proportion of the total number of particles.
Not a blocker for this PR, but I would like to discuss this default with you @ciguaran, I think there is evidence 0.5 is a better default value. in theory 0.8 should provide a similar result but with higher computational cost